contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -3435,3 +3510,415 @@ objectproperty_internal(PG_FUNCTION_ARGS)
+			if (catname == NULL)
+				ereport(ERROR,
+						(errcode(ERRCODE_SYNTAX_ERROR),
+						 errmsg("provided \"categories\" SQL must " \
+								"not return NULL values")));
+			
+			SPIcontext = MemoryContextSwitchTo(per_query_ctx);
+			catdesc = (bbf_pivot_cat_desc *) palloc(sizeof(bbf_pivot_cat_desc));
+			catdesc->catname = catname;
+			catdesc->attidx = i;
+			/* Add the proc description block to the hashtable */
+			bbf_pivot_HashTableInsert(bbf_pivot_hash, catdesc);
+
+			MemoryContextSwitchTo(SPIcontext);
+		}
+	}
+
+	if (SPI_finish() != SPI_OK_FINISH)
+		/* internal error */
+		elog(ERROR, "load_categories_hash: SPI_finish() failed");
+
+	return bbf_pivot_hash;
+}
+
+
+
+/*
+ * create and populate the bbf_pivot tuplestore
+ */
+static Tuplestorestate *
+get_bbf_pivot_tuplestore(RawStmt *sql,
+						HTAB *bbf_pivot_hash,
+						TupleDesc tupdesc,
+						bool randomAccess)
+{
+	Tuplestorestate *tupstore;
+	int			num_categories = hash_get_num_entries(bbf_pivot_hash);
+	AttInMetadata *attinmeta = TupleDescGetAttInMetadata(tupdesc);
+	char	  **values;
+	HeapTuple	tuple;
+	int			ret;
+	uint64		proc;
+
+	/* initialize our tuplestore (while still in query context!) */
+	tupstore = tuplestore_begin_heap(randomAccess, false, work_mem);
+
+	/* Connect to SPI manager */
+	if ((ret = SPI_connect()) < 0)
+		/* internal error */
+		elog(ERROR, "get_bbf_pivot_tuplestore: SPI_connect returned %d", ret);
+
+	/* Now retrieve the bbf_pivot source rows */
+	ret = SPI_execute_raw_parsetree(sql, true, 0);
+	proc = SPI_processed;
+
+	/* Check for qualifying tuples */
+	if ((ret == SPI_OK_SELECT) && (proc > 0))
+	{
+		SPITupleTable *spi_tuptable = SPI_tuptable;
+		TupleDesc	spi_tupdesc = spi_tuptable->tupdesc;
+		int			ncols = spi_tupdesc->natts;
+		char	   **columngroup;
+		char	   **lastcolumngroup = NULL;
+		bool		firstpass = true;
+		uint64		i;
+		int			j;
+		int			result_ncols;
+
+		if (num_categories == 0)
+		{
+			/* no qualifying category tuples */
+			ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					 errmsg("provided \"categories\" SQL must " \
+							"return 1 column of at least one row")));
+		}
+
+		if (ncols < 2)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("invalid source data SQL statement"),
+					 errdetail("The provided SQL must return 2 " \
+							   " columns; category, and values.")));
+
+		result_ncols = (ncols - 2) + num_categories;
+
+		/* Recheck to make sure we tuple descriptor still looks reasonable */
+		if (tupdesc->natts != result_ncols)
+			ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					 errmsg("invalid return type"),
+					 errdetail("Query-specified return " \
+							   "tuple has %d columns but bbf_pivot " \
+							   "returns %d.", tupdesc->natts, result_ncols)));
+
+		/* allocate space and make sure it's clear */
+		values = (char **) palloc0(result_ncols * sizeof(char *));
+		columngroup = (char **) palloc0((ncols - 2) * sizeof(char *));
```

Comment:
I see a bunch of `ncols-2`. Put this in a variable and use that variable instead and add a comment why the -2.